lnd.go

```diff
@@ -789,14 +792,94 @@ func waitForWalletPassword(grpcEndpoints, restEndpoints []string,
 		"Use `lncli create` to create wallet, or " +
 		"`lncli unlock` to unlock already created wallet.")
 
-	// We currently don't distinguish between getting a password to
-	// be used for creation or unlocking, as a new wallet db will be
-	// created if none exists when creating the chain control.
+	// We currently don't distinguish between getting a password to be used
+	// for creation or unlocking, as a new wallet db will be created if
+	// none exists when creating the chain control.
 	select {
-	case walletPw := <-pwService.CreatePasswords:
-		return walletPw, walletPw, nil
+
+	// The wallet is being created for the first time, we'll check to see
+	// if the user provided any entropy for seed creation. If so, then
+	// we'll create the wallet early to load the seed.
+	case initMsg := <-pwService.InitMsgs:
+		password := initMsg.Passphrase
+
+		// We'll now check to see fit eh user provided any entropy
+		// themselves. If not, then we'll generate the recommended
+		// number of random bytes to create an HD seed.
+		var (
+			generatedEntropy bool
+			entropy          [walletunlocker.SeedEntropySize]byte
+		)
+		if initMsg.SeedEntropy == walletunlocker.NoEntropySeed {
+			rpcsLog.Infof("No user entropy provided for wallet. " +
+				"Generating entropy...")
+
+			_, err := rand.Read(entropy[:])
+			if err != nil {
+				return nil, nil, err
+			}
+
+			generatedEntropy = true
+		} else {
+			rpcsLog.Infof("User entropy provided for wallet " +
+				"creation! Converting into HD seed...")
+			copy(entropy[:], initMsg.SeedEntropy[:])
+		}
+
+		// Before we can create the wallet, we'll first convert the
+		// 256-bits of entropy to the final seed using the BIP 39
+		// process.
+		mnemonic, err := bip39.NewMnemonic(entropy[:])
+		if err != nil {
+			return nil, nil, fmt.Errorf("unable to convert entropy "+
+				"to mnemonic: %v", err)
+		}
+
+		// Once we have the mnemonic we'll then convert from the
+		// mnemonic to the HD seed.
+		//
+		// TODO(roasbeef): don't use passphrase?
+		pass := string(password)
```

Comment:
It is kind of confusing to conflate the wallet encrypting `password` here with the BIP 39 passphrase. Seems like this needs to be an explicit passphrase, ideally supplied just after the 24-word mnemonic.